crates/fluvio/src/producer/dispatcher.rs

```diff
@@ -0,0 +1,305 @@
+use std::sync::Arc;
+use std::collections::HashMap;
+use futures_util::StreamExt;
+use flume::{Receiver, Sender};
+
+use tracing::{debug_span, debug, error, instrument};
+use tracing_futures::Instrument;
+use fluvio_types::event::StickyEvent;
+use dataplane::ReplicaKey;
+
+use crate::FluvioError;
+use crate::spu::SpuPool;
+use crate::producer::{Result, ProducerError};
+use crate::producer::{DispatcherMsg, ProducerConfig, ProducerMsg};
+use crate::producer::assoc::{AssociatedResponse, AssociatedRequest, AssociatedRecord, BatchStatus};
+use crate::producer::buffer::RecordBuffer;
+
+pub(crate) struct Dispatcher {
```

Comment:
Making this into a sequential controller  seems hard to reason about behavior.  For example, what if there is error in one of the producer.  Also, difficult to pipeline.   Rather, make each PartitionProducer as independent controller each with its' independent state machine.

Reply:
I think that is something we can certainly do for a v2, but if possible I would really like to get this merged since it is tested and working. I think the structure of the code is also good enough that we can update this in the future when we need to.